tools/wavetool.py

```diff
@@ -0,0 +1,150 @@
+#! /usr/bin/python3
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright(c) 2020 Intel Corporation. All rights reserved.
+
+"""
+A Tool to Generate and Manipulate Wave Files
+"""
+
+import os
+import sys
+import argparse
+import numpy as np
+import scipy.signal as signal
+import scipy.io.wavfile as wavefile
+
+def generate_sinusoids(**p):
+    """
+    Generate sine or cosine wave.
+
+    ``y(t) = A * sin(2 * pi * f * t + phi)``
+
+    ``y(t) = A * cos(2 * pi * f * t + phi)``
+
+    Parameters
+    ----------
+    dict of sinusiod parameters:
+    p['type']: Wave type, ``sine`` or ``cosine``
+    p['amp']: Amplitude
+    p['freq']: Frequency
+    p['phase']: Phase
+    p['chan']: Channels
+    p['sample_rate']: Sample rate
+    p['duration']: Duration
+
+    Returns
+    ----------
+    Real n-D Array with shape ``(duration * sample_rate, channel)``
+    """
+    func = np.sin if p['type'] == 'sine' else np.cos
+    time = np.arange(0, p['duration'], 1.0 / p['sample_rate'], dtype=np.float32)
+    data = p['amp'] * func(2 * np.pi * p['freq'] * time + p['phase'])
+    return np.reshape(np.repeat(data, p['chan']),[len(data), p['chan']])
+
+def generate_wav(**cmd):
+    if cmd['generate'] in ['sine', 'cosine']:
+        wave_param = {
+            'type': cmd['generate'], 'amp': cmd['amp'][0], 'freq': cmd['freq'][0],
+            'phase': cmd['phase'][0], 'chan': cmd['channel'], 'sample_rate': cmd['sample_rate'],
+            'duration': cmd['duration'][0]
+        }
+        wave_data = generate_sinusoids(**wave_param)
+    else:
+        raise Exception('invalid generate function, will generate nothing')
+    return wave_data
+
+def save_wave(wave_data, **cmd):
+    wave_path = cmd['output']
+    if wave_path == '.' or not wave_path.endswith('wav'):
+        wave_path = wave_path + '/tmp.wav'
+    sample_bits = cmd['bits']
+    np_types = {'S8': np.int8, 'S16': np.int16, 'S32': np.int32}
+    if sample_bits in np_types.keys():
+        # range of wave_data is floating point [-1.0, 1.0]. When saving to any integer format use its full range.
+        wave_data = (np.iinfo(np_types[sample_bits]).max * wave_data).astype(np_types[sample_bits])
+    wavefile.write(wave_path, cmd['sample_rate'], wave_data)
+
+def do_wave_analysis(**cmd):
+    fs_wav, wave = wavefile.read(cmd['recorded_wave'])
+    fs_ref, ref_wave = wavefile.read(cmd['reference_wave'])
+    if fs_wav != fs_ref:
+        print('Can not compare wave with different sample rate')
+        sys.exit(1)
+    if cmd['analyze'] == 'smart_amp':
+        analyze_wav_smart_amp(wave, ref_wave, fs_wav)
+
+# remove digital zeros in two sides
+def trim_wave(wave):
+    # once waves go through DAC/ADC, zero will become small value close to zero,
+    # here we set the digital zero threshold to 100, and cut samples below 100
+    # of two sides, this has no negative impact to processing.
+    DIGITAL_ZERO_THRESHOLD = 100
```

Comment:
Not a good idea to refer directly to integer values of s16/s24/s32 format. The code will fail with different word length used. Such threshold should be a float or decibel number relative to full scale.

But the most important question - why do you need to trim the samples based on this kind of silence detection?

Reply:
For DSM, there are only digital process, a comparison with zero is OK to trim zeros in two sides. but for WOV, to trim zero, we need to find a zero threshold because of ADC/DAC.

I have done some experiment with the current mounted DMIC injector under S16_LE,  and found 100 is a proper value to discard silence. I understand that under S24/S32 some more valid values may be trimmed, but it doesn't impact later data process that much.
Without much knowledge in audio processing senario, I found this strategy worked. But as you have stated, dBFS will provide the same criteria under s16/s24/s32, and is better to be used here. 
Will change to dBFS implementation in next push.

Reply:
If 100 is suitable for s16, then please change it to 100.0/INT16_MAX that is about 0.003. Or -50.3 decibels. I think that would be the minimum fix for this.

It's not possible to test e.g. volume control over it's range with such simple test signal find. It gets unreliable with low levels and noise. That's why I use in SOF/tools/test/audio stuff the chirp markers and code to detect them. It works with noisy ADC/DAC too and even with acoustics and some office noise. Lack of doing that is not blocker for smart amp test apparently but once implemented as library it's easy to use in all tests and make them robust.

Reply:
current implementation is OK to cope with the TGL release, later we can add more feature to this tool, for example, I'd like to add EQ test by checking if some frequency component is decreased.